src/main/java/com/google/cloud/discotoproto3converter/proto3/DocumentToProtoConverter.java

```diff
@@ -308,35 +308,35 @@ private void readResources(Document document) {
         String httpOptionPath = method.flatPath();
         List<String> methodSignatureParams = new ArrayList<>();
 
-        for (Schema requiredParam : method.requiredParams().values()) {
-          Field requiredField = schemaToField(requiredParam);
-          Option fieldBehaviorOption = new Option("google.api.field_behavior");
-          fieldBehaviorOption.getProperties().put("", "REQUIRED");
-          requiredField.getOptions().add(fieldBehaviorOption);
-          input.getFields().add(requiredField);
-          methodSignatureParams.add(requiredField.getName());
-        }
-
         for (Schema pathParam : method.pathParams().values()) {
           Field pathField = schemaToField(pathParam);
+          if (method.requiredParamNames().contains(pathParam.getIdentifier())) {
+            pathField.getOptions().add(getFieldBehaviorOption());
+          }
+          input.getFields().add(pathField);
+          methodSignatureParams.add(pathField.getName());
           httpOptionPath =
               httpOptionPath.replace(
                   "{" + pathParam.getIdentifier() + "}", "{" + pathField.getName() + "}");
         }
 
         Option requiredMethodSignatureOption = new Option("google.api.method_signature");
-        requiredMethodSignatureOption.getProperties().put("", String.join(",", methodSignatureParams));
+        requiredMethodSignatureOption
+            .getProperties()
+            .put("", String.join(",", methodSignatureParams));
 
         for (Schema queryParam : method.queryParams().values()) {
           Field queryField = schemaToField(queryParam);
+          if (method.requiredParamNames().contains(queryParam.getIdentifier())) {
```

Comment:
Are we adding this required `queryParam` to the `requiredMethodSignatureOption`?

And what about required fields in the body of the message?

Reply:
Nice catch, that was a bug. After looking closer to the whole logic I had to make a significant change to it.

Now it works as follows:
1) The `"parameterOrder"` method property in disocovery json defines the list of parameters to go to method_signature and their order.
2) If a method has a request object, also include it in the method_signature as the last argument (i.e. message body is always a required parameter).

This is a pretty simple and reliable logic which should satisfy most use cases and is resistant to backward-incompatible changes.